### DISCUSSION TITLE
Dynamic setting of service instance ID

### DISCUSSION THREAD

**Discussion by srl100** [COVESA/vsomeip#760] [Q&A]
I am wondering whether the vsomeip & CommonAPI code can be used to create a service with a dynamic instance ID.

The use case for this is that I want to run the same executable file on several servers and have them offer different instances of the same service, as per SOME/IP's PRS_SOMEIP_00162 requirement.

For example, two servers on my network with IP addresses 192.168.1.1 and 192.168.1.2 should be able to offer the same service 0x1234 but with different instance IDs (0x0101 and 0x0102).

It seems to me that the SomeIpInstanceID will be baked in to the code generated, as it is present in the .fdepl file and raises an error if omitted:

```
define org.genivi.commonapi.someip.deployment for provider Service {
    instance commonapi.examples.E01HelloWorld {
        InstanceId = "commonapi.examples.HelloWorld"
        
        SomeIpInstanceID = 22136
    
        SomeIpUnicastAddress = "192.168.0.2"
        SomeIpReliableUnicastPort = 30499
        SomeIpUnreliableUnicastPort = 30499
    }
}
```
and it seems from experiment that if this instance ID does not match that of one of the services in the proxy's .json file, then the instance is not offered by the routing manager:

(file: CommonAPI-Examples/E01HelloWorld/vsomeip-service.json) 
```
            "services" :
            [
                {
                    "service" : "0x1234",
                    "instance" : "0x5678",
                    "reliable" : { "port" : "30509", "enable-magic-cookies" : "false" },
                    "unreliable" : "31000"
                }
            ]
```

NB: 0x5678 = 22136 decimal

I can manipulate the contents of the .json file before starting the proxy, if required, but I can't recompile the .fdepl file on-target!

**Comment by tomas1284** [comment]
Note that it's possible to provide **multiple** someip deployments with different SomeIpInstanceID's in **a single fdepl file**:
```
define org.genivi.commonapi.someip.deployment for provider Service {
    instance commonapi.examples.E01HelloWorld {
        InstanceId = "commonapi.examples.HelloWorld.instance01"        
        SomeIpInstanceID = 22136
    }
}
define org.genivi.commonapi.someip.deployment for provider Service {
    instance commonapi.examples.E01HelloWorld {
        InstanceId = "commonapi.examples.HelloWorld.instance02"
        SomeIpInstanceID = 22137
    }
}
```

This will produce proxy and stub which are aware of both instances. Afterwards, your services or clients should then use the corresponding instance when registering the service or building the proxy, respectively, e.g.:

```
std::string instance;
if (useFirstInstance)
    instance = "commonapi.examples.HelloWorld.instance01";
else
    instance = "commonapi.examples.HelloWorld.instance02";

runtime->registerService(domain, instance, myService, connection);
```

**Reply by srl100** [reply]
Thanks @tomas1284 - that answers the specific case in my example (of having two servers) but does not really scale in an elegant way to the more general case of having N possible servers as we end up with N x "_define ... deployment for provider Service_" sections in the .fdepl file!

My real-world use case is for an ISO 17215 compliant video camera using DHCP - part of specifcation ISO 17215-3:2021 says "_InstanceID should be set to the last octet of the camera's IP address_", which would give us 256 service provider definitions in one .fdepl file!
